content/blog/how-to-analyze-grpc-with-wireshark.md

```diff
@@ -0,0 +1,157 @@
+
+## Features of Wireshark Protobuf and gRPC dissectors
+
+The main features about Wireshark Protobuf and gRPC dissectors includes:
+
+- Supports dissecting gRPC messages serialized in Protobuf or JSON format.
+
+- Supports dissecting the gRPC messages of unary, server streaming, client streaming, and bidirectional streaming RPC calls.
+
+- Message and service definition language files (*.proto) can be configured to enable more precise parsing of the serialized data in Protobuf format for protocols such as gRPC.
+
+- Protobuf fields can be dissected as Wireshark fields that allows user to input the full name of a Protobuf field or message in the filter toolbar to search packages containing the field or message.
+
+- You can register your own subdissectors in the 'protobuf_field' dissector table, which is keyed with the full names of the fields, for further parsing Protobuf fields of BYTES or STRING type.
+
+- You can add the mapping record between UDP port and Protobuf message type in the Protobuf protocol preferences dialog, so that the Wireshark can decode the traffic on that UDP port as this Protobuf message.
+
+- You can create your own dissectors to parsing the protocols based on Protobuf over UDP or TCP.
+
+## Capture gRPC traffic
+
+There are many ways to capture the network traffic of a gRPC conversation:
+
+- Windows: Use Wireshark to capture traffic from many different network media types, including Ethernet, Wireless LAN, and more.
+
+- Linux/UNIX: Use Tcpdump.
+
+Please refer to [Wireshark User's Guide]https://www.wireshark.org/docs/wsug_html_chunked/ for how to capture network packets files that can be recognized by Wireshark.
+
+Note that now only the capture files that sending gRPC messages in plaintext mode can be parsed by Wireshark. For example, you have to setup a gRPC client channel using plaintext mode in java like:
+```java
+NettyChannelBuilder.forAddress(host, port).negotiationType(NegotiationType.PLAINTEXT).build();
+```
+
+In theory, as long as the session secret keys of TLS can be exported in format of Wireshark [key log file](https://gitlab.com/wireshark/wireshark/-/wikis/tls) from client or server side, the gRPC traffic encrypted in TLS can also be parsed by Wireshark. But as I know, gRPC stub libraries do not support exporting session secret keys of TLS now or not support directly. Hope this feature might be supported in the future.
```

Comment:
@ejona86 @mjpitz @jtattermusch - given the limitations mentioned in lines 40-45 (copied below for your convenience), how useful is this really? (Emphasis mine)

> Note that now **only** the capture files that sending gRPC messages in **plaintext mode** can be parsed by Wireshark. ...
>
> **In theory**, as long as the session secret keys of TLS can be exported in format of Wireshark [key log file](https://gitlab.com/wireshark/wireshark/-/wikis/tls) from client or server side, the gRPC traffic encrypted in TLS can also be parsed by Wireshark. But as I know, **gRPC stub libraries do not support exporting session secret keys** of TLS now or not support directly. Hope this feature might be supported in the future.

So, in theory, this is usable, but in practice it isn't 😄 ?

Reply:
Netty supports to export session secret Keys of tls https://github.com/netty/netty/blob/master/handler/src/main/java/io/netty/handler/ssl/SslMasterKeyHandler.java as wireshark key log file format.
But it seems grpc-java does not merge it yet.

Reply:
Thanks for the clarifications. I'll wait for @ejona86 @mjpitz @jtattermusch to comment, which most likely won't be until next week.

Reply:
I find there might have been some progress on it (https://github.com/grpc/grpc-java/issues/7199).

Reply:
Plaintext is (unfortunately) pretty common and many times when using Wireshark you can make it plaintext when you need to. I do think plain-text is enough for the blog post. This raises awareness that things like proto-inspection are even available, and common paths can become smoother.

SSLKEYLOGFILE may be functional today in C; it doesn't appear to be in Java, although I didn't investigate why. Both use boringssl which seems to have support; maybe we need a special flag. It looks there's a way to get the keys today [with one line of Go TLS configuration](https://golang.org/pkg/crypto/tls/#example_Config_keyLogWriter).

Many people aren't using Wireshark, so nobody's worked to make keys available. If we say "no point in using Wireshark because keys are available" then it becomes a chicken and egg problem.

Reply:
That makes sense. Thanks for the thoughtful feedback @ejona86.